### PR TITLE
Syntactic Marshaller

### PR DESCRIPTION
**Pull request by adlrocha**
# Goal
- [x] marshal: syntactic IR (literals + dict) —> json
- [x] marshal: json —> syntactic IR (literals + dict)
   - [x] String
   - [x] Blob
   - [x] Dict
   - [x] Pairs
   - [x] Number (Int)
   - [x] Number (Float)
   - [x] Bool

- [x] tests
- [ ] Simple benchmark?

### CODE REVIEW

**Comment by adlrocha** [comment]
@petar I think this is ready for review. If you it is OK with you I would merge this to your branch so we keep there the most up to date version of the code. I will open another branch for the VM work I will start now.

File: ir/blob.go [hunk 1]
```diff
@@ -16,6 +17,40 @@ func (b Blob) WritePretty(w io.Writer) error {
 	return err
 }
 
+func (b Blob) MarshalJSON() (bdata []byte, e error) {
+	// Temporal type to avoid recursion
+	type tmp Blob
```
Comment (petar) [inline ir/blob.go:22]: This is fine. I just don't see why you couldn't use type []byte instead of introducing a new type tmp.
(Same for other places you do this.)

File: ir/dict.go [hunk 2]
```diff
@@ -142,6 +164,27 @@ func (d Dict) Get(key Node) Node {
 	return nil
 }
 
+func (d Dict) MarshalJSON() (b []byte, e error) {
+	// Temporal type to avoid recursion
+	type tmp Dict
+	ts := tmp(d)
+
+	return json.Marshal(&ts)
```
Comment (petar) [inline ir/dict.go:172]: Does this work? Where are you writing the "type" and "value" fields?

File: ir/ir.go [hunk 3]
```diff
@@ -6,4 +6,6 @@ import (
 
 type Node interface {
 	WritePretty(w io.Writer) error
+	//UnmarshalJSON(data []byte) error
+	MarshalJSON() (b []byte, e error)
```
Comment (petar) [inline ir/ir.go:10]: use "json.Marshaler" here

File: ir/marshal.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	"encoding/json"
+	"fmt"
+	"io"
+)
+
+type MarshalType int
+
+// List of types
+const (
+	StringType = iota
```
Comment (petar) [inline ir/marshal.go:13]: Use strings, not (implicitly generated) numbers. It's easier for forward compatibility.

File: ir/marshal_test.go [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+	byteData := b.Bytes()
+
+	// Decode
+	r := bytes.NewReader(byteData)
+	out := Dict{}
+	err = Unmarshal(r, &out)
+	if err != nil {
+		t.Fatal(err)
+	}
+
```
Comment (petar) [inline ir/marshal_test.go:36]: There is no test here. You need to add:
if !IsEqual(n, out) { ... }

File: ir/marshal_test.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	no := Pairs{}
+	// Encode
+	valueBytes, err := json.Marshal(n)
+	if err != nil {
+		panic(err)
+	}
+	err = json.Unmarshal(valueBytes, &no)
+	if err != nil {
+		panic(err)
+	}
```
Comment (petar) [inline ir/marshal_test.go:165]: No test here too.

File: ir/numbers.go [hunk 7]
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
...
+
+	}
+	return nil
+}
+
+func (n Int) MarshalJSON() (b []byte, e error) {
+	c := struct {
+		Type  MarshalType `json:"type"`
+		Value string      `json:"value"`
+	}{Type: IntType, Value: n.String()}
```
Comment (petar) [inline ir/numbers.go:136]: I thought we said the Value field here will equal the output of big.Int.MarshalText().

File: ir/numbers.go [hunk 8]
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
...
+	return json.Marshal(&c)
+}
+
+func (n Float) MarshalJSON() (b []byte, e error) {
+	c := struct {
+		Type  MarshalType `json:"type"`
+		Value string      `json:"value"`
+	}{Type: FloatType, Value: n.String()}
```
Comment (petar) [inline ir/numbers.go:144]: Same as above.

File: ir/numbers.go [hunk 9]
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
...
+}
+
+func unmarshalFloat(p []byte, n *Float) error {
+	if string(p) == "null" {
+		return nil
+	}
+	// Using precision 53 to accommodate for big.NewFloat which uses 53 preciosion by default.
+	// If we change this marshalling behavior to use MarshalText we need to change to 64 precision.
+	// Check: https://github.com/golang/go/issues/45309
+	z, _, err := big.ParseFloat(strings.Replace(string(p), "\"", "", -1), 10, 53, big.ToNearestEven)
```
Comment (petar) [inline ir/numbers.go:81]: Here, we said we would use big.Float.UnmarshalText (the inverse of big.Float.MarshalText)

File: ir/numbers.go [hunk 10]
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
 	}
 	panic("bug: unknown number type")
 }
+
+func unmarshalInt(p []byte, n *Int) error {
+	if string(p) == "null" {
+		return nil
+	}
+	var z big.Int
+	_, ok := z.SetString(strings.Replace(string(p), "\"", "", -1), 10)
```
Comment (petar) [inline ir/numbers.go:66]: Same as below: use big.Int.UnmarshalText

File: ir/marshal.go [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+	BlobType
+	PeerType
+	IntType
+	FloatType
+	BoolType
+)
+
+// UnmarshalType list of supported types
+func UnmarshalType(tp MarshalType, b []byte) (Node, error) {
```
Comment (petar) [inline ir/marshal.go:22]: Document what is b. Is this the encoding of the type/value-wrapper, or is it the encoding just of the value? It looks like it is just the encoding of the value (no wrapper).

File: ir/numbers.go [hunk 12]
```diff
@@ -54,3 +57,90 @@ func IsEqualNumber(x, y Number) bool {
...
+func (n *Float) UnmarshalJSON(b []byte) error {
+	var objMap map[string]*json.RawMessage
+	err := json.Unmarshal(b, &objMap)
+	if err != nil {
+		err = unmarshalFloat(b, n)
+
+		if err != nil {
+			return err
+		}
+	} else {
```
Comment (petar) [inline ir/numbers.go:97]: This method should only ever be called by "UnmarshalType(tp MarshalType, b []byte) (Node, error)", which passes the encoding of the value in the b parameter. What is the purpose of the code in the "else" block below?

I think I know what the issue is. You are trying to make your top-level user-facing method
  func Unmarshal(r io.Reader, out Node) error
to work, when someone passes any concrete syntactic type for "out". 

This causes these "ugly" if/else statements here. The ugliness is actually the fact that you are checking for types in two different places in your code: once in every concrete type's unmarshal method, and once in the UnmarshalType method. This means that there is no one source of truth in the code and makes the code very prone to future bugs.

To fix this issue, change the top-level unmarshaling interface to:
func Unmarshal(r io.Reader) (Node, error)

File: ir/blob.go (hunk 1 above)
Comment (adlrocha) [reply ir/blob.go:22]: I tried, you end up inside an infinite loop because when trying to Marshal a few lines below Go enters this same function. I can share a few funny Go playgrounds with this behavior :)

Comment (petar) [reply ir/blob.go:22]: What I mean is:
func (b Blob) MarshalJSON() (bdata []byte, e error) {
  c := struct {
		Type  MarshalType `json:"type"`
		Value []byte         `json:"value"`
	}{Type: BlobType, Value: b.Bytes}
	return json.Marshal(&c)
}

File: ir/dict.go (hunk 2 above)
Comment (adlrocha) [reply ir/dict.go:172]: This works "out-of-the-box". The fact that we implemented the MarshalJSON interface for every type allows json.Marshal to know how to Marshal Pairs without additional information. It works with all the tests that I've done, but I may be missing something.

File: ir/blob.go (hunk 1 above)
Comment (adlrocha) [reply ir/blob.go:22]: Oh, I see! Doing so breaks the unmarshalling of Pairs, because json.Unmarshal doesn't know what type to use. The fact that we use String{Value: string}, Blob{Value: []byte{}} makes this easy for these types. And this is why I need to use the "ugly" if-else in Float and Int, to accommodate for the fact that we don't use the internal "Value" in number types.

However, I suspect all of this is linked to the top-level interface thing you mention in your last comment.

File: ir/numbers.go (hunk 12 above)
Comment (adlrocha) [reply ir/numbers.go:93]: I'd say that I only check types in one place, when using UnmarshalJSON for Pairs. For the rest of types the json.Marshaller knows how to unmarshal the right type by itself, am I right?

And regarding the top-level interface, I did it this way because I wanted to use Decode under the hood, and I didn't know how to write a MarshalJSON for Node. In any case, would you mind if we discuss this sync? I think I may be missing something here. Thanks!

**Comment by adlrocha** [comment]
Thank you such a quick review, @petar. I've addressed every comment except a few things regarding code structure and the high-level unmarshal interface where I have some questions (as you'll see in the comments). Ping me if we can have quick sync chat to solve them.

File: ir/dict.go (hunk 2 above)
Comment (adlrocha) [reply ir/dict.go:178]: You were right here! I was missing the case where a `Dict` is part of another `Dict` inside a `Pair. I completely missed that case and that's why I thought it worked. Apologies! 

Will push the new version in a bit (we have a few meetings now and I want to revise it first).

Comment (adlrocha) [reply ir/dict.go]: Now it is working. The test fails though because of this https://github.com/libp2p/go-smart-record/issues/5

Comment (adlrocha) [reply ir/dict.go]: Tests pass. The issue was that in the test I had to pairs with the same `Key`

**Comment by adlrocha** [comment]
@petar, I finished the refactor addressing all your comments from yesterday and the tests now pass. Let me know if there is anything else to do before merging. Thanks!

File: ir/marshal.go [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+	FloatType  = "Float"
+	BoolType   = "Bool"
+	DictType   = "Dict"
+	SetType    = "Set"
+)
+
+// decodeMultiplex does the unmarshalling of the type
+// once the wrapper has been process and the Node type has
+// been identified.
+func decodeMultiplex(v interface{}) (Node, error) {
```
Comment (petar) [inline ir/marshal.go:24]: knit: lets call this decodeNode.

File: ir/base/cid.go [hunk 14]
```diff
@@ -25,3 +26,7 @@ func (c Cid) WritePretty(w io.Writer) error {
 func (c Cid) MergeWith(ctx ir.MergeContext, x ir.Node) (ir.Node, error) {
 	panic("XXX")
 }
+
```
Comment (petar) [inline ir/base/cid.go:29]: why is this needed? Smart interface does not call for UnmarshalJSON method.

File: ir/bool.go [hunk 15]
```diff
@@ -17,3 +17,14 @@ func (b Bool) WritePretty(w io.Writer) (err error) {
...
+
+func (b Bool) encodeJSON() (interface{}, error) {
+	return struct {
+		Type  marshalType `json:"type"`
+		Value bool        `json:"value"`
+	}{Type: BoolType, Value: b.Value}, nil
+}
+
+func decodeBool(s map[string]interface{}) (Node, error) {
+	return Bool{s["value"].(bool)}, nil
```
[the remark this answers is not in the document]
Comment (petar) [reply ir/bool.go:29]: Same comment applies everywhere (there are many places in decodeXXX, where you type-cast without a check).

**Review by petar — COMMENTED** [review]
Very nice. We are almost there. A few small comments.

File: ir/base/cid.go (hunk 14 above)
Comment (adlrocha) [reply ir/base/cid.go:29]: It is not! Forgot to clean it from some tinkering I was doing. Thanks!

File: ir/dict.go [hunk 16]
```diff
@@ -25,6 +25,8 @@ func (p Pair) WritePretty(w io.Writer) error {
 	return nil
 }
 
+func (p Pair) encodeJSON() (interface{}, error) { return nil, nil }
```
[the remark this answers is not in the document]
Comment (adlrocha) [reply ir/dict.go]: I had to add it because we were returning a `Pair` as a node in `dict.Remove`. I changed that function to return a boolean instead so we don't have to Pair as a node.

File: ir/dict.go [hunk 17]
```diff
@@ -139,16 +140,15 @@ func (d Dict) CopySetTag(tag string, key Node, value Node) Dict {
 	return c
 }
 
-func (d *Dict) Remove(key Node) Node {
+func (d *Dict) Remove(key Node) bool {
```
Comment (adlrocha) [inline ir/dict.go:143]: Changing returning value to prevent `Pair` from having to be a `Node`. We could also return directly a `Pair` instead of a `Node` if we want to return the removed `Pair`.

**Comment by adlrocha** [comment]
Thank you for the review, @petar. One last comment and I think we are ready to merge.

File: ir/dict.go (hunk 17 above)
Comment (petar) [reply ir/dict.go:143]: This is my mistake in the original code. What should be returned is the deleted Value (which is a Node), or nil if nothing was deleted.

**Review by petar — APPROVED** [review]
One small change, otherwise LGTM.
